db/badgerdb/db.go

```diff
@@ -0,0 +1,402 @@
+)
+
+var (
+	versionsFilename string = "versions.csv"
+)
+
+var _ dbm.DBConnection = (*BadgerDB)(nil)
+var _ dbm.DBReader = (*badgerTxn)(nil)
+var _ dbm.DBWriter = (*badgerWriter)(nil)
+var _ dbm.DBReadWriter = (*badgerWriter)(nil)
+
+// BadgerDB is a connection to a BadgerDB key-value database.
+type BadgerDB struct {
+	db   *badger.DB
+	vmgr *versionManager
+	mtx  sync.RWMutex
+	// Track open DBWriters
+	openWriters int32
+}
+
+type badgerTxn struct {
+	txn *badger.Txn
+	// vmgr *versionManager
+	db *BadgerDB
+}
+type badgerWriter struct {
+	badgerTxn
+}
+
+type badgerIterator struct {
+	reverse    bool
+	start, end []byte
+	iter       *badger.Iterator
+	lastErr    error
+	primed     bool
+}
+
+// Map our versions to Badger timestamps.
+//
+// A badger Txn's commit TS must be strictly greater than a record's "last-read"
+// TS in order to detect conflicts, and a Txn must be read at a TS after last
+// commit to see current state. So we must use commit increments that are more
+// granular than a version interval, mapping latter -> former.
+type versionManager struct {
+	*dbm.VersionManager
+	vmap   versionTsMap
+	lastTs uint64
+}
+type versionTsMap map[uint64]uint64
+
+// NewDB creates or loads a BadgerDB key-value database inside the given directory.
+// If dir does not exist, it will be created.
+func NewDB(dir string) (*BadgerDB, error) {
+	// Since Badger doesn't support database names, we join both to obtain
+	// the final directory to use for the database.
+	if err := os.MkdirAll(dir, 0755); err != nil {
+		return nil, err
+	}
+	opts := badger.DefaultOptions(dir)
+	// todo: NumVersionsToKeep
+	opts.SyncWrites = false // note that we have Sync methods
+	opts.Logger = nil       // badger is too chatty by default
+	return NewDBWithOptions(opts)
+}
+
+// NewDBWithOptions creates a BadgerDB key-value database with the specified Options
+// (https://pkg.go.dev/github.com/dgraph-io/badger/v3#Options)
+func NewDBWithOptions(opts badger.Options) (*BadgerDB, error) {
+	db, err := badger.OpenManaged(opts)
+	if err != nil {
+		return nil, err
+	}
+	vmgr, err := readVersionsFile(filepath.Join(opts.Dir, versionsFilename))
+	if err != nil {
+		return nil, err
+	}
+	return &BadgerDB{
+		db:   db,
+		vmgr: vmgr,
+	}, nil
+}
+
+// Load metadata CSV file containing valid versions
+func readVersionsFile(path string) (*versionManager, error) {
+	file, err := os.OpenFile(path, os.O_RDONLY|os.O_CREATE, 0644)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+	r := csv.NewReader(file)
+	r.FieldsPerRecord = 2
+	rows, err := r.ReadAll()
+	if err != nil {
+		return nil, err
+	}
+	var versions []uint64
+	var ts uint64
+	vmap := make(versionTsMap)
+	for _, row := range rows {
+		version, err := strconv.ParseUint(row[0], 10, 64)
+		if err != nil {
+			return nil, err
+		}
+		ts, err = strconv.ParseUint(row[1], 10, 64)
+		if err != nil {
+			return nil, err
+		}
+		versions = append(versions, version)
+		vmap[version] = ts
+	}
+	if ts == 0 {
+		ts = 1
+	}
```

Comment:
Can you add a comment for this code block pls?

Reply:
Actually, this can be dropped. It was there to handle the empty case, and ensure the max timestamp is > 0, but it's not needed.
This made me notice a different bug, though - this code and the versionmanager constructor should not require sorted input. Will fix.